guardian/editorial-tools-user-telemetry-service
# Add authentication to the API lambda's `event` endpoint

## What does this change?

This PR adds pan-domain authentication to the API lambda's `event` endpoint.

It also adds the scaffolding necessary to run the lambda in a local environment – a setup script in `scripts/setup`, and an nginx mapping to route requests from `user-telemetry.local.dev-gutools.co.uk`. This is necessary to exercise pan-domain authentication in a local context.

It mocks authentication in tests, so authentication behaviour is manually tested for now. I've tested in local and PROD environments.

## How to test

Run `./scripts/setup` locally to set up the local domain name, and run the local lambda following the project instructions (see `projects/event-lambdas`. Requests to the `event` endpoint should fail without a panda cookie. 

You can acquire a panda cookie by running another project that signs panda cookies (e.g. Composer) and copying the cookie from that app's request headers.

<details>
<summary>Example curl request</summary>

```
curl 'https://user-telemetry.local.dev-gutools.co.uk/event' \
  -H 'Content-Type: application/json' \
  -H 'Cookie: <Cookie body goes here>' \
  -X POST \
  -d '[{"app":"example-app","stage":"PROD","type":"USER_ACTION_1","value":1,"eventTime":"2020-09-03T07:51:27.669Z"},{"app":"example-app","stage":"PROD","type":"USER_ACTION_2","value":1,"eventTime":"2020-09-03T07:51:27.669Z"}]'
```
</details>

## How can we measure success?

Users with authenticated cookies are able to post events to our telemetry service. Everyone/thing else cannot.

## Have we considered potential risks?

We do not authenticate the other endpoint exposed here, `/healthcheck`, which simply issues a 200 response, and does not expose any other functionality.
